core/engine/src/builtins/date/utils.rs

```diff
@@ -750,22 +753,37 @@ pub(super) fn pad_six(t: u32, output: &mut [u8; 6]) -> JsStr<'_> {
 /// [spec-format]: https://tc39.es/ecma262/#sec-date-time-string-format
 pub(super) fn parse_date(date: &JsString, hooks: &dyn HostHooks) -> Option<i64> {
     // All characters must be ASCII so we can return early if we find a non-ASCII character.
-    let Ok(date) = date.to_std_string() else {
-        return None;
+    let owned_js_str = date.as_str();
+    let owned_string: String;
+    let date = match owned_js_str.variant() {
+        JsStrVariant::Latin1(s) => {
+            if !s.is_ascii() {
+                return None;
+            }
+            // SAFETY: Since all characters are ASCII we can safely convert this into str.
+            unsafe { str::from_utf8_unchecked(s) }
+        }
+        JsStrVariant::Utf16(s) => {
+            owned_string = String::from_utf16(s).ok()?;
+            if !owned_string.is_ascii() {
+                return None;
+            }
+            owned_string.as_str()
+        }
```

Comment:
Funnily enough, I think this could probably use `Cow` to avoid creating an `owned_string` variable that is only used once.
```suggestion
    let date = match owned_js_str.variant() {
        JsStrVariant::Latin1(s) => {
            if !s.is_ascii() {
                return None;
            }
            // SAFETY: Since all characters are ASCII we can safely convert this into str.
            Cow::Borrowed(unsafe { str::from_utf8_unchecked(s) })
        }
        JsStrVariant::Utf16(s) => {
            date = String::from_utf16(s).ok()?;
            if !owned_string.is_ascii() {
                return None;
            }
            Cow::Owned(date)
        }
```